Test: build images

achilleas-k opened the pull request:
New `cmd/` that builds an image based on a config (blueprint + image options).  Distro and image type selectionis handled by command line arguments.
Configs have been created in `test/configs/` based on the `format-request-map.json`.
A new gitlab job creates new jobs dynamically, one for each build.

The generated config currently looks like this:
```yaml
build/fedora-38/qcow2:
    stage: test
    script: go run ./cmd/build -distro "fedora-38" -image "qcow2" -config ./test/configs/empty.json
    extends: .terraform
    variables:
        RUNNER: aws/fedora-38-x86_64
        INTERNAL_NETWORK: "false"

build/fedora-38/ami:
    stage: test
    script: go run ./cmd/build -distro "fedora-38" -image "ami" -config ./test/configs/empty.json
    extends: .terraform
    variables:
        RUNNER: aws/fedora-38-x86_64
        INTERNAL_NETWORK: "false"

build/centos-9/qcow2:
    stage: test
    script: go run ./cmd/build -distro "centos-9" -image "qcow2" -config ./test/configs/empty.json
    extends: .terraform
    variables:
        RUNNER: aws/fedora-38-x86_64
        INTERNAL_NETWORK: "false"

build/centos-9/ami:
    stage: test
    script: go run ./cmd/build -distro "centos-9" -image "ami" -config ./test/configs/empty.json
    extends: .terraform
    variables:
        RUNNER: aws/fedora-38-x86_64
        INTERNAL_NETWORK: "false"
```

Limiting the build for now to test the config.

achilleas-k commented on test/cases/generate-build-config.sh:
```diff
@@ -0,0 +1,65 @@
+build/${distro}/${imgtype}:
+    stage: test
+    script: |
+        sudo dnf install -y go gpgme-devel gcc osbuild osbuild-luks2 osbuild-lvm2 osbuild-ostree osbuild-selinux
```
I'll borrow the osbuild pinning setup from composer but I think by default (when there's no pins) we should install the latest version from the repos like any other dependency right?

As for the runners, fully agree.  I'll set some up tomorrow.

ondrejbudai replied:
ftr, @croissanne is working on some automation for building runner images automatically and with deps installed. :) Might be worth syncing with her.

achilleas-k commented:
Let's try to test everything!

achilleas-k commented:
Found a bug where we would ignore osbuild errors if the output was unstructured.

supakeen commented on test/cases/build-image.sh:
```diff
@@ -0,0 +1,55 @@
+
+arch=$(uname -m)
+
+# TODO: Include osbuild commit hash
```
For now you could include `osbuild --version` first so any large discrepancies show up.

achilleas-k replied:
Done!

supakeen commented on test/cases/generate-build-config:
```diff
@@ -0,0 +1,233 @@
+    generation will be incorrect.
+    """
+    bad_configs = []
+    for fname in os.listdir(CONFIGS_PATH):
+        fpath = os.path.join(CONFIGS_PATH, fname)
+        with open(fpath) as config_file:
+            config = json.load(config_file)
+            config_filename, _ = os.path.splitext(fname)
+            if config_filename != config["name"]:
+                bad_configs.append(fpath)
```
I prefer:
```python
for file in pathlib.Path(CONFIGS_PATH).glob("*.json"):
   config = json.loads(file.read_text())
   if file.stem != config["name"]:
       bad_configs.append(str(file))
```

achilleas-k replied:
Done!

supakeen reviewed (commented):
Not yet complete but I took a partial look.

supakeen commented:
By the way the output in CI is <3

achilleas-k commented:
Found another bug: The way the manifest is marshalled in `./cmd/build` will escape some characters like `>` and create essentially a different manifest.  This is why some images are always being rebuilt now.  They use shell redirection in some blueprint options and those get escaped so the manifests never match when checking if they need to be rebuilt.

achilleas-k commented:
Fixed yet another bug that I created when adding the osbuild version to the build cache.  I fixed the bug and also added error handling to the config generator.

supakeen reviewed (approved):
I've gone through the code, the Go and Python code look good.

Let's merge it and find bugs.
